Modules/Server/Sources/PocketCastsServer/Public/Sync/SyncYearListeningHistoryTask.swift

```diff
@@ -0,0 +1,158 @@
+                performRequest(token: token, shouldSync: true)
+            } else {
+                success = true
+            }
+        } catch {
+            print("SyncYearListeningHistory had issues decoding protobuf \(error.localizedDescription)")
+        }
+    }
+
+    private func syncMissingEpisodes(serverData: Data) {
+        do {
+            let response = try Api_YearHistoryResponse(serializedData: serverData)
+
+            // on watchOS, we don't show history, so we also don't process server changes we only want to push changes up, not down
+            #if !os(watchOS)
+            updateEpisodes(updates: response.history.changes)
+            #endif
+
+            success = true
+        } catch {
+            print("SyncYearListeningHistory had issues decoding protobuf \(error.localizedDescription)")
+        }
+    }
+
+    private func updateEpisodes(updates: [Api_HistoryChange]) {
+        var podcastsToUpdate: [String] = []
+
+        let dispatchGroup = DispatchGroup()
+
+        for change in updates {
+            dispatchGroup.enter()
+
+            DispatchQueue.global(qos: .userInitiated).async {
+                let interactionDate = Date(timeIntervalSince1970: TimeInterval(change.modifiedAt / 1000))
+
+                let episodeInDatabase = DataManager.sharedManager.findEpisode(uuid: change.episode)
+
+                if episodeInDatabase == nil {
+                    // Episode is not on database, let's add it
+
+                    ServerPodcastManager.shared.addMissingPodcastAndEpisode(episodeUuid: change.episode, podcastUuid: change.podcast)
+                    DataManager.sharedManager.setEpisodePlaybackInteractionDate(interactionDate: interactionDate, episodeUuid: change.episode)
+                    podcastsToUpdate.append(change.podcast)
+                } else if episodeInDatabase?.lastPlaybackInteractionDate == nil {
+                    // Episode is in database but it's not updated, let's update it
+
+                    DataManager.sharedManager.setEpisodePlaybackInteractionDate(interactionDate: interactionDate, episodeUuid: change.episode)
+                    podcastsToUpdate.append(change.podcast)
+                }
+
+                dispatchGroup.leave()
+            }
+        }
+
+        dispatchGroup.wait()
+
+        // Sync episode status for the retrieved podcasts' episodes
+        let uniqueUuidsToUpdate = Array(Set(podcastsToUpdate))
+        updateEpisodes(for: uniqueUuidsToUpdate)
+    }
+
+    private func updateEpisodes(for podcastsUuids: [String]) {
+        podcastsUuids.forEach {
+            if let episodes = ApiServerHandler.shared.retrieveEpisodeTaskSynchronouusly(podcastUuid: $0) {
+                DataManager.sharedManager.saveBulkEpisodeSyncInfo(episodes: DataConverter.convert(syncInfoEpisodes: episodes))
+            }
+        }
+    }
+}
+
+/// Helper that checks for podcast existence
+/// It caches database requests
+class PodcastExistHelper {
```

Comment:
I created this to avoid calling the database over and over to check if an episode exists.